crates/bevy_ui/src/layout_components.rs

```diff
@@ -0,0 +1,430 @@
+    Component,
+    Deref,
+    DerefMut,
+    Copy,
+    Clone,
+    PartialEq,
+    Debug,
+    Default,
+    Serialize,
+    Deserialize,
+    Reflect,
+)]
+#[reflect_value(PartialEq, Serialize, Deserialize)]
+pub struct Offset(pub UiRect);
+
+/// Controls the size of UI nodes
+///
+/// Layout is performed according to the [`Layout`]
+/// To check the actual size of a UI element, read its [`Transform`](bevy_transform::components::Transform) component
+#[derive(Component, Copy, Clone, PartialEq, Debug, Default, Serialize, Deserialize, Reflect)]
+#[reflect_value(PartialEq, Serialize, Deserialize)]
+pub struct SizeConstraints {
+    /// The minimum extent, which cannot be violated by the layouting algorithm
+    pub min: Size,
+    /// The suggested extent, which will be used if other constraints can be comfortably satisfied
+    pub suggested: Size,
+    /// The maximum extent, which cannot be violated by the layouting algorithm
+    pub max: Size,
+    /// The expected aspect ratio, computed as width / height
+    pub aspect_ratio: Option<f32>,
+}
+
+impl SizeConstraints {
+    pub const DEFAULT: SizeConstraints = SizeConstraints {
+        min: Size::DEFAULT,
+        suggested: Size::DEFAULT,
+        max: Size::DEFAULT,
+        aspect_ratio: None,
+    };
+
+    pub const FULL: SizeConstraints = SizeConstraints {
+        min: Size::DEFAULT,
+        suggested: Size::FULL,
+        max: Size::FULL,
+        aspect_ratio: None,
+    };
+
+    /// Sets only the minimum extent
+    pub const fn min(width: Val, height: Val) -> SizeConstraints {
+        SizeConstraints {
+            min: Size::new(width, height),
+            ..Self::DEFAULT
+        }
+    }
+
+    /// Sets only the suggested extent
+    pub const fn suggested(width: Val, height: Val) -> SizeConstraints {
+        SizeConstraints {
+            suggested: Size::new(width, height),
+            ..Self::DEFAULT
+        }
+    }
+
+    /// Sets only the suggested extent
+    pub const fn max(width: Val, height: Val) -> SizeConstraints {
+        SizeConstraints {
+            max: Size::new(width, height),
+            ..Self::DEFAULT
+        }
+    }
+}
+
+/// The space around and inside of a UI node
+#[derive(Component, Copy, Clone, PartialEq, Debug, Default, Serialize, Deserialize, Reflect)]
+#[reflect_value(PartialEq, Serialize, Deserialize)]
+pub struct Spacing {
+    /// The space around the outside of the UI element
+    pub margin: UiRect,
+    /// The space around the inside of the UI element
+    pub padding: UiRect,
+    /// The space around the outside of the UI element that can be colored to create a visible border
+    pub border: UiRect,
+}
+
+impl Spacing {
+    pub const DEFAULT: Spacing = Spacing {
+        margin: UiRect::DEFAULT,
+        padding: UiRect::DEFAULT,
+        border: UiRect::DEFAULT,
+    };
+
+    /// Adds an automatic margin on all sides.
+    ///
+    /// When the element is explicitly sized, this will (hopefully) cause the element
```

Comment:
Using `hopefully` in a doc makes me really not trust the api.

Reply:
That's okay, I think that suspicion is warranted. As @plof27 explained it to me, this is a classic CSS flexbox hack that has all sorts of hard to explain and predict edge cases but mostly works!

Reply:
This is indeed a CSS trick, but it's a rather historical one widely used before flex box came out. Furthermore, it works only horizontally, and people usually need a table cell hack (wrapping with a table and a cell, then set the cell alignment to vertical) to align vertically. In CSS world, auto-margin trick works for `display: block` in both `position: static` and `position: relative` but not `position: absolute`.

For CSS flex box, to horizontally center an item we use `justify-content: center` to center along flex box's flowing axis, then use `align-items: center` to center items on the cross axis (usually vertical).

I'm not too sure the use case of this, is this centering designed to work with `position: relative`? I'm not too sure if replicating these tricks could be a good idea. Personally, I'm a fan of SwiftUI's idea of having explicit `Spacing` item in their hstack/vstack hierarchy.

Reply:
> this is a classic CSS flexbox hack 

If that's the case, I would rather not have it hard coded as a const in Bevy. Or if it's kept, to not mention that part in the comment

Reply:
We should not be advocating for using old hacks if there is a correct way to do it.

Reply:
Mhmm, I'm in agreement. The pattern came up repeatedly in the old examples, and I was confused about what its intent was.